version 4.3.3 breaks TS/Angular/webpack

App that compiles and runs using immutable 4.3.2, fails to start after upgrading to 4.3.3. Seems to be caused by https://github.com/immutable-js/immutable-js/pull/1932

Attempting to run app results in error:
`Error: Failed to initialize Angular compilation - Maximum call stack size exceeded`

To duplicate:

1. Create an angular app - `npx @angular/cli new TestApp` (choose no routing and css)
2. Install immutable 4.3.2 - `npm i immutable@4.3.2`
3. In tsconfig.json change the value of `noImplicitOverride` option from true to false.
4. Replace the contents of app.component.ts with the following (which kind of captures the problematic code in more or less the way its actually presenting):
```
import { Component } from '@angular/core';
import { /*DeepCopy,*/ List, Record } from 'immutable';

export interface Props {
  foos?: List<Foo>;
}

const defaults: Props = {
  foos: undefined // null
};

const fooRecord = Record(defaults, 'Foo');

export class Foo extends fooRecord implements Props {
  foos?: List<Foo>;
}

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent {
  title = 'TestApp';

  // either of these causes issue
  // bar: DeepCopy<Foo>;
  bar = List<Foo>([new Foo()]).toJS();
}
```
5. Run the app - `npm start`. It starts as expected.
6. Upgrade to immutable 4.3.3 - `npm i immutable@4.3.3`
7. Run the app - `npm start`. App fails to start with `Error: Failed to initialize Angular compilation - Maximum call stack size exceeded`

If line 15 - `foos?: List<Foo>;` is commented out we get a more descriptive error:
```
Error: src/app/app.component.ts:28:3 - error TS2615: Type of property 'foos' circularly references itself in mapped type '{ [ObjectKey in keyof Props]: DeepCopy<Props[ObjectKey]>; }'.
28   bar = List<Foo>([new Foo()]).toJS();
```

--- a/type-definitions/immutable.d.ts
+++ b/type-definitions/immutable.d.ts
@@ -91,6 +91,16 @@
  */
 
 declare namespace Immutable {
+  /** @ignore */
+  type OnlyObject<T> = Extract<T, object>;
+
+  /** @ignore */
+  type ContainObject<T> = OnlyObject<T> extends object
+    ? OnlyObject<T> extends never
+      ? false
+      : true
+    : false;
+
   /**
    * @ignore
    *
@@ -100,14 +110,14 @@ declare namespace Immutable {
   export type DeepCopy<T> = T extends Record<infer R>
     ? // convert Record to DeepCopy plain JS object
       {
-        [key in keyof R]: DeepCopy<R[key]>;
+        [key in keyof R]: ContainObject<R[key]> extends true ? unknown : R[key];
       }
     : T extends Collection.Keyed<infer KeyedKey, infer V>
     ? // convert KeyedCollection to DeepCopy plain JS object
       {
         [key in KeyedKey extends string | number | symbol
           ? KeyedKey
-          : string]: DeepCopy<V>;
+          : string]: V extends object ? unknown : V;
       }
     : // convert IndexedCollection or Immutable.Set to DeepCopy plain JS array
     T extends Collection<infer _, infer V>
@@ -118,7 +128,9 @@ declare namespace Immutable {
     ? Array<DeepCopy<V>>
     : T extends object // plain JS object are converted deeply
     ? {
-        [ObjectKey in keyof T]: DeepCopy<T[ObjectKey]>;
+        [ObjectKey in keyof T]: ContainObject<T[ObjectKey]> extends true
+          ? unknown
+          : T[ObjectKey];
       }
     : // other case : should be kept as is
       T;

--- a/type-definitions/ts-tests/deepCopy.ts
+++ b/type-definitions/ts-tests/deepCopy.ts
@@ -23,10 +23,12 @@ import { List, Map, Record, Set, Seq, DeepCopy, Collection } from 'immutable';
   // $ExpectType { [x: string]: string; }
   type StringKey = DeepCopy<Map<string, string>>;
 
-  // $ExpectType { [x: string]: object; }
+  // should be `{ [x: string]: object; }` but there is an issue with circular references
+  // $ExpectType { [x: string]: unknown; }
   type ObjectKey = DeepCopy<Map<object, object>>;
 
-  // $ExpectType { [x: string]: object; [x: number]: object; }
+  // should be `{ [x: string]: object; [x: number]: object; }` but there is an issue with circular references
+  // $ExpectType { [x: string]: unknown; [x: number]: unknown; }
   type MixedKey = DeepCopy<Map<object | number, object>>;
 
   // $ExpectType string[]
@@ -55,10 +57,12 @@ import { List, Map, Record, Set, Seq, DeepCopy, Collection } from 'immutable';
 {
   // Nested
 
-  // $ExpectType { map: { [x: string]: string; }; list: string[]; set: string[]; }
+  // should be `{ map: { [x: string]: string; }; list: string[]; set: string[]; }` but there is an issue with circular references
+  // $ExpectType { map: unknown; list: unknown; set: unknown; }
   type NestedObject = DeepCopy<{ map: Map<string, string>; list: List<string>; set: Set<string>; }>;
 
-  // $ExpectType { map: { [x: string]: string; }; }
+  // should be `{ map: { [x: string]: string; }; }`, but there is an issue with circular references
+  // $ExpectType { map: unknown; }
   type NestedMap = DeepCopy<Map<'map', Map<string, string>>>;
 }
 
@@ -68,6 +72,51 @@ import { List, Map, Record, Set, Seq, DeepCopy, Collection } from 'immutable';
   type Article = Record<{ title: string; tag: Tag; }>;
   type Tag = Record<{ name: string; article: Article; }>;
 
-  // $ExpectType { title: string; tag: { name: string; article: any; }; }
+  // should handle circular references here somehow
+  // $ExpectType { title: string; tag: unknown; }
   type Circular = DeepCopy<Article>;
+}
+
+{
+  // Circular references #1957
+
+  class Foo1 extends Record<{
+    foo: undefined | Foo1;
+  }>({
+    foo: undefined
+  }) {
+  }
+
+  class Foo2 extends Record<{
+    foo?: Foo2;
+  }>({
+    foo: undefined
+  }) {
+  }
+
+  class Foo3 extends Record<{
+    foo: null | Foo3;
+  }>({
+    foo: null
+  }) {
+  }
+
+  // $ExpectType { foo: unknown; }
+  type DeepFoo1 = DeepCopy<Foo1>;
+
+  // $ExpectType { foo?: unknown; }
+  type DeepFoo2 = DeepCopy<Foo2>;
+
+  // $ExpectType { foo: unknown; }
+  type DeepFoo3 = DeepCopy<Foo3>;
+
+  class FooWithList extends Record<{
+    foos: undefined | List<FooWithList>;
+  }>({
+    foos: undefined
+  }) {
+  }
+
+  // $ExpectType { foos: unknown; }
+  type DeepFooList = DeepCopy<FooWithList>;
 }

--- a/type-definitions/ts-tests/record.ts
+++ b/type-definitions/ts-tests/record.ts
@@ -88,7 +88,8 @@ import { List, Map, Record, RecordOf, Set } from 'immutable';
   // $ExpectType { map: Map<string, string>; list: List<string>; set: Set<string>; }
   withMap.toJSON();
 
-  // $ExpectType { map: { [x: string]: string; }; list: string[]; set: string[]; }
+  // should be `{ map: { [x: string]: string; }; list: string[]; set: string[]; }` but there is an issue with circular references
+  // $ExpectType { map: unknown; list: unknown; set: unknown; }
   withMap.toJS();
 }
 
@@ -101,7 +102,8 @@ import { List, Map, Record, RecordOf, Set } from 'immutable';
 
   const line = Line({});
 
-  // $ExpectType { size?: { distance: string; } | undefined; color?: string | undefined; }
+  // should be  { size?: { distance: string; } | undefined; color?: string | undefined; } but there is an issue with circular references
+  // $ExpectType { size?: unknown; color?: string | undefined; }
   line.toJS();
 }
 

